File: app/src/main/java/com/chessclub/app/ui/game/SubmitGameActivity.java
package com.chessclub.app.ui.game;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;

import com.chessclub.app.R;
import com.chessclub.app.database.DatabaseHelper;
import com.chessclub.app.database.GameDao;
import com.chessclub.app.database.PlayerDao;
import com.chessclub.app.model.Game;
import com.chessclub.app.model.Player;

import java.util.ArrayList;
import java.util.List;

/**
 * Activity for submitting new chess games
 */
public class SubmitGameActivity extends AppCompatActivity {
    private static final String PREF_NAME = "ChessClubPrefs";
    private static final String KEY_LOGGED_IN_USER_ID = "loggedInUserId";

    private Toolbar toolbar;
    private Spinner spinnerWhitePlayer;
    private Spinner spinnerBlackPlayer;
    private RadioGroup radioGroupResult;
    private Button btnSubmit;
    private TextView tvError;
    
    private PlayerDao playerDao;
    private GameDao gameDao;
    private List<Player> players;
    private int whitePlayerId = -1;
    private int blackPlayerId = -1;
    private int currentPlayerId = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_submit_game);
        
        // Initialize database
        DatabaseHelper dbHelper = DatabaseHelper.getInstance(this);
        playerDao = dbHelper.getPlayerDao();
        gameDao = dbHelper.getGameDao();
        
        // Get current user ID
        SharedPreferences prefs = getSharedPreferences(PREF_NAME, MODE_PRIVATE);
        currentPlayerId = prefs.getInt(KEY_LOGGED_IN_USER_ID, -1);
        
        // Initialize views
        toolbar = findViewById(R.id.toolbar);
        spinnerWhitePlayer = findViewById(R.id.spinner_white_player);
        spinnerBlackPlayer = findViewById(R.id.spinner_black_player);
        radioGroupResult = findViewById(R.id.radio_group_result);
        btnSubmit = findViewById(R.id.btn_submit);
        tvError = findViewById(R.id.tv_error);
        
        // Set up toolbar
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setTitle("Submit Game");
        
        // Load players for spinners
        loadPlayers();
        
        // Set up spinner listeners
        setupSpinnerListeners();
        
        // Set up submit button
        btnSubmit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                submitGame();
            }
        });
    }

    /**
     * Load players from database and populate spinners
     */
    private void loadPlayers() {
        players = playerDao.getAllPlayers();
        
        if (players.isEmpty()) {
            showError("No players found in the system");
            btnSubmit.setEnabled(false);
            return;
        }
        
        // Create adapters for spinners
        List<String> playerNames = new ArrayList<>();
        for (Player player : players) {
            playerNames.add(player.getName() + " (" + player.getElo() + ")");
        }
        
        ArrayAdapter<String> adapter = new ArrayAdapter<>(
                this, android.R.layout.simple_spinner_item, playerNames);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        
        spinnerWhitePlayer.setAdapter(adapter);
        spinnerBlackPlayer.setAdapter(adapter);
        
        // Try to set current user as white player by default
        if (currentPlayerId != -1) {
            for (int i = 0; i < players.size(); i++) {
                if (players.get(i).getId() == currentPlayerId) {
                    spinnerWhitePlayer.setSelection(i);
                    break;
                }
            }
        }
        
        // Set opponent as black player (next player in list)
        if (players.size() > 1) {
            int nextIndex = (spinnerWhitePlayer.getSelectedItemPosition() + 1) % players.size();
            spinnerBlackPlayer.setSelection(nextIndex);
        }
    }

    /**
     * Set up listeners for player selection spinners
     */
    private void setupSpinnerListeners() {
        spinnerWhitePlayer.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                whitePlayerId = players.get(position).getId();
                validatePlayerSelection();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                whitePlayerId = -1;
                validatePlayerSelection();
            }
        });

        spinnerBlackPlayer.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                blackPlayerId = players.get(position).getId();
                validatePlayerSelection();
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                blackPlayerId = -1;
                validatePlayerSelection();
            }
        });
    }

    /**
     * Validate player selection
     */
    private void validatePlayerSelection() {
        if (whitePlayerId == blackPlayerId && whitePlayerId != -1) {
            showError("White and black players cannot be the same");
            btnSubmit.setEnabled(false);
        } else {
            hideError();
            btnSubmit.setEnabled(true);
        }
    }

    /**
     * Submit game to database
     */
    private void submitGame() {
        if (whitePlayerId == -1 || blackPlayerId == -1) {
            showError("Please select both players");
            return;
        }
        
        if (whitePlayerId == blackPlayerId) {
            showError("White and black players cannot be the same");
            return;
        }
        
        int selectedRadioId = radioGroupResult.getCheckedRadioButtonId();
        if (selectedRadioId == -1) {
            showError("Please select a game result");
            return;
        }
        
        // Determine result
        Game.Result result;
        if (selectedRadioId == R.id.radio_white_win) {
            result = Game.Result.WHITE_WIN;
        } else if (selectedRadioId == R.id.radio_black_win) {
            result = Game.Result.BLACK_WIN;
        } else {
            result = Game.Result.DRAW;
        }
        
        // Record game in database
        long gameId = gameDao.recordGameWithEloUpdate(whitePlayerId, blackPlayerId, result);
        
        if (gameId > 0) {
            // Show success message
            Toast.makeText(this, "Game recorded successfully", Toast.LENGTH_SHORT).show();
            finish(); // Go back to previous screen
        } else {
            showError("Failed to record game. Please try again.");
        }
    }

    /**
     * Display error message
     * @param message Error message to display
     */
    private void showError(String message) {
        tvError.setText(message);
        tvError.setVisibility(View.VISIBLE);
    }

    /**
     * Hide error message
     */
    private void hideError() {
        tvError.setVisibility(View.GONE);
    }

    @Override
    public boolean onSupportNavigateUp() {
        onBackPressed();
        return true;
    }
}

File: ChessClubAndroid/app/src/main/java/com/chessclub/android/ui/game/SubmitGameActivity.kt
package com.chessclub.android.ui.game

import android.os.Bundle
import android.view.View
import android.widget.ArrayAdapter
import android.widget.AutoCompleteTextView
import android.widget.Button
import android.widget.EditText
import android.widget.ProgressBar
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.chessclub.android.R
import com.chessclub.android.model.Player

class SubmitGameActivity : AppCompatActivity() {
    private lateinit var viewModel: SubmitGameViewModel
    
    private lateinit var whitePlayerAutoComplete: AutoCompleteTextView
    private lateinit var blackPlayerAutoComplete: AutoCompleteTextView
    private lateinit var whitePlayerPinEdit: EditText
    private lateinit var blackPlayerPinEdit: EditText
    private lateinit var resultRadioGroup: RadioGroup
    private lateinit var submitButton: Button
    private lateinit var cancelButton: Button
    private lateinit var progressBar: ProgressBar
    
    private var players: List<Player> = listOf()
    private var playerMap: Map<String, String> = mapOf() // Maps player name to ID
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_submit_game)
        
        viewModel = ViewModelProvider(this).get(SubmitGameViewModel::class.java)
        
        // Initialize views
        whitePlayerAutoComplete = findViewById(R.id.white_player_autocomplete)
        blackPlayerAutoComplete = findViewById(R.id.black_player_autocomplete)
        whitePlayerPinEdit = findViewById(R.id.white_player_pin)
        blackPlayerPinEdit = findViewById(R.id.black_player_pin)
        resultRadioGroup = findViewById(R.id.result_radio_group)
        submitButton = findViewById(R.id.submit_button)
        cancelButton = findViewById(R.id.cancel_button)
        progressBar = findViewById(R.id.progress_bar)
        
        // Set up observers
        observeViewModel()
        
        // Load player data
        viewModel.loadPlayers()
        
        // Set up button actions
        submitButton.setOnClickListener { submitGame() }
        cancelButton.setOnClickListener { finish() }
    }
    
    private fun observeViewModel() {
        viewModel.players.observe(this) { playerList ->
            players = playerList
            
            // Create map for player name to ID lookup
            playerMap = playerList.associateBy({ it.name }, { it.id })
            
            // Set up autocomplete adapters
            val playerNames = playerList.map { it.name }
            val adapter = ArrayAdapter(this, android.R.layout.simple_dropdown_item_1line, playerNames)
            
            whitePlayerAutoComplete.setAdapter(adapter)
            blackPlayerAutoComplete.setAdapter(adapter)
        }
        
        viewModel.isLoading.observe(this) { isLoading ->
            progressBar.visibility = if (isLoading) View.VISIBLE else View.GONE
            submitButton.isEnabled = !isLoading
        }
        
        viewModel.gameCreated.observe(this) { success ->
            if (success) {
                Toast.makeText(this, getString(R.string.game_submitted), Toast.LENGTH_SHORT).show()
                finish()
            } else {
                Toast.makeText(this, getString(R.string.error_submitting_game), Toast.LENGTH_LONG).show()
            }
        }
    }
    
    private fun submitGame() {
        val whitePlayerName = whitePlayerAutoComplete.text.toString()
        val blackPlayerName = blackPlayerAutoComplete.text.toString()
        val whitePlayerPin = whitePlayerPinEdit.text.toString()
        val blackPlayerPin = blackPlayerPinEdit.text.toString()
        
        // Validate inputs
        if (whitePlayerName.isBlank() || blackPlayerName.isBlank() || 
            whitePlayerPin.isBlank() || blackPlayerPin.isBlank()) {
            Toast.makeText(this, getString(R.string.fill_all_fields), Toast.LENGTH_SHORT).show()
            return
        }
        
        if (whitePlayerName == blackPlayerName) {
            Toast.makeText(this, getString(R.string.different_players_required), Toast.LENGTH_SHORT).show()
            return
        }
        
        // Get selected result
        val resultRadioId = resultRadioGroup.checkedRadioButtonId
        if (resultRadioId == -1) {
            Toast.makeText(this, getString(R.string.select_result), Toast.LENGTH_SHORT).show()
            return
        }
        
        val resultRadioButton = findViewById<RadioButton>(resultRadioId)
        val gameResult = when (resultRadioButton.id) {
            R.id.radio_white_wins -> "1-0"
            R.id.radio_black_wins -> "0-1"
            R.id.radio_draw -> "1/2-1/2"
            else -> ""
        }
        
        // Get player IDs
        val whitePlayerId = playerMap[whitePlayerName]
        val blackPlayerId = playerMap[blackPlayerName]
        
        if (whitePlayerId == null || blackPlayerId == null) {
            Toast.makeText(this, getString(R.string.invalid_player_names), Toast.LENGTH_SHORT).show()
            return
        }
        
        // Submit game
        viewModel.submitGame(
            whitePlayerId,
            blackPlayerId,
            whitePlayerPin,
            blackPlayerPin,
            gameResult
        )
    }
}
